src/psyclone/tests/psyir/transformations/replace_reference_by_literal_trans_test.py

```diff
@@ -0,0 +1,282 @@
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+# COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: H. Brunie, University of Grenoble Alpes
+
+
+"""This module tests the ReplaceReferenceByLiteralTrans transformation.
+"""
+
+import pytest
+
+from psyclone.psyir.nodes import Literal, Routine
+from psyclone.psyir.symbols import INTEGER_TYPE
+from psyclone.psyir.transformations import (
+    ReplaceReferenceByLiteralTrans,
+    TransformationError,
+)
+
+
+# ----------------------------------------------------------------------------
+def test_rrbl_general():
+    """Test general functionality of the transformation."""
+
+    rrbl = ReplaceReferenceByLiteralTrans()
+
+    assert (
+        str(rrbl)
+        == "Replaces all static const Reference "
+        + "by its Literal in a subroutine."
+    )
+    assert rrbl.name == "ReplaceReferenceByLiteralTrans"
+
+
+# ----------------------------------------------------------------------------
+def test_rrbl_errors():
+    """Test errors that should be thrown."""
+
+    rrbl = ReplaceReferenceByLiteralTrans()
+    lit = Literal("1", INTEGER_TYPE)
+    with pytest.raises(TransformationError) as err:
+        rrbl.apply(lit)
+
+    assert (
+        "Error in ReplaceReferenceByLiteralTrans transformation. The "
+        "supplied node argument should be a PSyIR Routine, but found "
+        "'Literal'" in str(err.value)
+    )
+
+
+# ----------------------------------------------------------------------------
+def test_rrbl_working(fortran_reader, fortran_writer):
+    """Tests if subroutine parameters are replaced as expected."""
+    source = """program test
+                use mymod
+                type(my_type):: t1, t2, t3, t4
+                integer, parameter :: x=3, y=12, z=13
+                integer, parameter :: u1=1, u2=2, u3=3, u4=4
+
+                integer i, invariant, ic1, ic2, ic3
+                real, dimension(10) :: a
+                invariant = 1
+                do i = 1, 10
+                    t1%a = z
+                    a(ic1) = u1+(ic1+x)*ic1
+                    a(ic2) = u2+(ic2+y)*ic2
+                    a(ic3) = u3+(ic3+z)*ic3
+                    a(t1%a) = u4+(t1%a+u4*z)*t1%a
```

Comment:
I'm sure it will be fine but please could you extend this test to cover more edge cases:
  * constant var in a loop bound, e.g. `do i = u1, u4, u1`
  * constant var as index in derived-type access, e.g. `t1%b(u4)`
  * constant var as argument to subroutine call, e.g. `call my_sub(ic1+u3)`